facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+// Oauth token used to impersonate another service account.
+struct OauthTokenProvider {
+    service_account_to_impersonate: Option<String>,
+    default_service_account_oauth_token: Option<OauthToken>,
+    impersonated_service_account_oauth_token: Option<OauthToken>,
+}
+
+impl OauthTokenProvider {
+    // Creates a token provider which can impersonate the specified service
+    // account.
+    fn new(service_account_to_impersonate: Option<String>) -> OauthTokenProvider {
+        OauthTokenProvider {
+            service_account_to_impersonate,
+            default_service_account_oauth_token: None,
+            impersonated_service_account_oauth_token: None,
+        }
+    }
+
+    fn ensure_storage_access_oauth_token(&mut self) -> Result<String> {
+        match self.service_account_to_impersonate {
+            Some(_) => self.ensure_impersonated_service_account_oauth_token(),
+            None => self.ensure_default_service_account_oauth_token(),
+        }
+    }
+
+    // Obtains an Oauth token for the default service account, if one is not
+    // already held or if it has expired. Otherwise returns a copy of the token.
+    // The returned value is an owned reference because the token owned by this
+    // struct could change while the caller is still holding the returned token.
+    fn ensure_default_service_account_oauth_token(&mut self) -> Result<String> {
+        match &self.default_service_account_oauth_token {
+            Some(token) if !token.expired() => Ok(token.token.clone()),
+            _ => {
+                let http_response = ureq::get(DEFAULT_OAUTH_TOKEN_URL)
+                    .set("Metadata-Flavor", "Google")
+                    // By default, ureq will wait forever to connect or
+                    // read.
+                    .timeout_connect(10_000) // ten seconds
+                    .timeout_read(10_000) // ten seconds
+                    .call();
+                if http_response.error() {
+                    return Err(anyhow!(
+                        "failed to query GKE metadata service: {:?}",
+                        http_response
+                    ));
+                }
+
+                let response = http_response
+                    .into_json_deserialize::<MetadataServiceTokenResponse>()
+                    .context("failed to deserialize response from GKE metadata service")?;
+
+                if response.token_type != "Bearer" {
+                    return Err(anyhow!("unexpected token type {}", response.token_type));
+                }
+
+                self.default_service_account_oauth_token = Some(OauthToken {
+                    token: response.access_token.clone(),
+                    expiration: Utc::now() + Duration::seconds(response.expires_in),
+                });
+
+                Ok(response.access_token)
+            }
+        }
+    }
+
+    // Obtains an Oauth token to impersonate the service account this provider
```

Comment:
Same comment as above; I think it's clearer to rewrite this in the "Returns X, and gets X if needed" form.

That said, it looks like this function doesn't actually have a short-circuit to return early if there's already an impersonation token.